Pathfinding Merge

JohnByrneJames opened the pull request:
Merge between feature/pathfinding with master.

JohnByrneJames committed:
Added A* pathfinding to the game

Made a test scene for A* tracking using Gizmo's adding movement in next addition.

JohnByrneJames committed:
Added A* Heap Optimisation and Path Weightings

A* now moves accurately towards target on a path request basis, the process happens quicker due to Heap Optimisation [halves processing]. Weighting adds element of choice for agent to decide whether they will rather use a road rather than muddy terrain ECT.

JohnByrneJames committed:
Extra Script - Didn't get added before (Bug)?

Added small change.

JohnByrneJames committed:
Merge remote-tracking branch 'origin/feature/pathfinding' into feature/pathfinding

JohnByrneJames committed:
Added weighting Smoothing - Stable

This is stable but needs changes in the future.

JohnByrneJames committed:
Tower Defence - Basic Test Scene

Working turret/ turret placement (NEED TO ADD MOBILE COMPATIBILITY) WIP

JohnByrneJames committed:
Tower Defence | Final Version for Demo

Game works and you can place turrets to kill enemies, currently no currency system or end of level but includes path finding concepts.

Commit:
Adding multiple camera functionality to level

Rokkii committed:
Delete Gyro.cs

Rokkii committed:
Delete EditorBuildSettings.asset

JohnByrneJames committed:
Add features of level progression

Waves can be configured in CodeMonkey - waves, amount, enemy, rate between ECT.

JohnByrneJames committed:
Prefab created of pathfinding map

JohnByrneJames committed:
Checking for conflicts

JohnByrneJames committed:
Merge remote-tracking branch 'origin/master' into feature/pathfinding

JohnByrneJames committed:
Merge remote-tracking branch 'origin/master' into feature/pathfinding

JohnByrneJames committed:
Merge remote-tracking branch 'origin/feature/pathfinding' into feature/pathfinding

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Enemy.cs:
```diff
@@ -5,10 +5,14 @@
 
 public class Enemy : MonoBehaviour
 {
+    public float startSpeed = 10f;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

JohnByrneJames replied:
I think as it is this is far too picky for the prototype? It works and surely doesn't reduce performance.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Enemy.cs:
```diff
@@ -5,10 +5,14 @@
 {
+    public float startSpeed = 10f;
+
+    [HideInInspector] public float speed;
```
Please make this `private` and remove the `HideInInspectorAttribute`. No reason for the attribute to exist. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name _inThisFormat, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Enemy.cs:
```diff
@@ -5,10 +5,14 @@
+
+    [HideInInspector] public float speed;
+    
     public float startHealth = 100;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Enemy.cs:
```diff
@@ -5,10 +5,14 @@
-    public int worth = 50;
+    //public int worth = 50;
 
     public GameObject deathEffect;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Enemy.cs:
```diff
@@ -5,10 +5,14 @@
+    [HideInInspector] public float speed;
+    
     public float startHealth = 100;
     private float health;
```
Please update the naming convention of this to `_health` to bring it in line with the rest of the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Enemy.cs:
```diff
@@ -5,10 +5,14 @@
     private float health;
 
-    public int worth = 50;
+    //public int worth = 50;
```
Dead code. Please delete.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Enemy.cs:
```diff
@@ -5,10 +5,14 @@
 
 public class Enemy : MonoBehaviour
```
It seems this entire type is breaking our conventions. I've pointed out what I can but some of the stuff is not part of this branch so I can't point it out. Please apply things that seem to apply in places I've not marked out. Thanks.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Enemy.cs:
```diff
@@ -19,6 +23,7 @@ public class Enemy : MonoBehaviour
 
     void Start ()
```
Missing access modifier. Please provide one.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+public class GameManager : MonoBehaviour
+{
+
+    public static bool GameIsOver;
```
`static` is strictly prohibited for use cases like this. It is brittle and not good design. Please redesign this to not be `static`. Thanks. Assuming this is public-facing data you might also want to consider this to be a property, or whatever fits your new design.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+    public GameObject completeLevelUI;
+
+    void start()
+    {
```
You can expression body this method.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+        if (GameIsOver)
+            return;
+        /*
+        if (PlayerStats.Lives <= 0)
```
Dead code. Please delete.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+    void Update()
+    {
+        if (GameIsOver)
+            return;
```
Please put this on the same line as the if statement.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+        GameIsOver = false;
+    }
+
+    void Update()
```
This isn't doing anything other than providing a performance hit. Is this even needed?

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    /*
+    void EndGame()
```
Dead code. Please delete.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+
+    public void WinLevel()
+    {
+        //GameIsOver = true;
```
Dead code. Please delete.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+
+    //public GameObject gameOverUI;
+
+    public GameObject completeLevelUI;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+
+    public static bool GameIsOver;
+
+    //public GameObject gameOverUI;
```
Dead code. Please delete.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/GameManager.cs:
```diff
@@ -0,0 +1,44 @@
+    }
+    */
+
+    public void WinLevel()
```
You can expression body this method.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/PlayerStats.cs:
```diff
@@ -0,0 +1,17 @@
+
+public class PlayerStats : MonoBehaviour
+{
+    //public static int Money; 
```
Dead code. Please delete.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/PlayerStats.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    //public static int Money; 
+    
+    // public static int Lives;
```
Dead code. Please delete.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/PlayerStats.cs:
```diff
@@ -0,0 +1,17 @@
+    
+    // public static int Lives;
+
+    public static int Rounds;
```
`static` is strictly prohibited for use cases like this. It is brittle and not good design. Please redesign this to not be `static`. Thanks. Assuming this is public-facing data you might also want to consider this to be a property, or whatever fits your new design.

JohnByrneJames replied:
If I remove this it doesn't work - if it works then its good for the prototype. Will think of a work around later

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/PlayerStats.cs:
```diff
@@ -0,0 +1,17 @@
+
+    public static int Rounds;
+
+    void Start()
```
Missing access modifier. Please provide an access modifier.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/PlayerStats.cs:
```diff
@@ -0,0 +1,17 @@
+    public static int Rounds;
+
+    void Start()
+    {
```
You can expression body this method.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/RoundsSurvived.cs:
```diff
@@ -0,0 +1,32 @@
+
+public class RoundsSurvived : MonoBehaviour
+{
+    public Text roundsText;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/RoundsSurvived.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    public Text roundsText;
+
+    void OnEnable()
```
Missing access modifier. Please provide an access modifier.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/RoundsSurvived.cs:
```diff
@@ -0,0 +1,32 @@
+    public Text roundsText;
+
+    void OnEnable()
+    {
```
You can expression body this method.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/RoundsSurvived.cs:
```diff
@@ -0,0 +1,32 @@
+        StartCoroutine(AnimateText());
+    }
+
+    IEnumerator AnimateText()
```
Missing access modifier. Please provide an access modifier.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/RoundsSurvived.cs:
```diff
@@ -0,0 +1,32 @@
+        roundsText.text = "0";
+        int round = 0;
+
+        yield return new WaitForSeconds(.7f);
```
This float value is badly formatted. Please provide full formatting.

RubyNova replied:
making new `WaitForSeconds` like this every time is expensive. Please create and persist the `WaitForSeconds` you require in private fields and simply `yield return` the fields.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/RoundsSurvived.cs:
```diff
@@ -0,0 +1,32 @@
+            round++;
+            roundsText.text = round.ToString();
+            
+            yield return new WaitForSeconds(.05f);
```
This float value is badly formatted. Please provide full formatting.

RubyNova replied:
making new `WaitForSeconds` like this every time is expensive. Please create and persist the `WaitForSeconds` you require in private fields and simply `yield return` the fields.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Wave.cs:
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+[System.Serializable]
```
Please add a `using System;` statement and make this a `SerializableAttribute` rather than `System.SerializeableAttribute`.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Wave.cs:
```diff
@@ -0,0 +1,11 @@
+[System.Serializable]
+public class Wave
+{
+    public GameObject enemy;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Wave.cs:
```diff
@@ -0,0 +1,11 @@
+public class Wave
+{
+    public GameObject enemy;
+    public int count;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Wave.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public GameObject enemy;
+    public int count;
+    public float rate;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
 public class WaveSpawner : MonoBehaviour
 {
-    public Transform enemyPrefab;
+    public static int EnemiesAlive = 0;
```
`static` is strictly prohibited for use cases like this. It is brittle and not good design. Please redesign this to not be `static`. Thanks. Assuming this is public-facing data you might also want to consider this to be a property, or whatever fits your new design.

JohnByrneJames replied:
tried turning the static into get and set but did not work

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
-    public Transform enemyPrefab;
+    public static int EnemiesAlive = 0;
+    
+    public Wave[] waves;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

JohnByrneJames replied:
It doesn't exist anymore.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
+    
+    public Wave[] waves;
 
     public Transform[] spawnPoints;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
 
     public Transform[] spawnPoints;
 
     public float timeBetweenWaves = 5f;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
-    private float initialCountdown = 2f;
+    private float countdown = 2f;
 
     public Text waveCountdownText;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
+    private float countdown = 2f;
 
     public Text waveCountdownText;
+    public Text enemiesRemaining;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
+    public Text enemiesRemaining;
 
-    private int waveIndex = 1;
+    public GameManager gameManager;
```
public fields are a cardinal sin and go against the convention of the codebase. If you need Unity inspector support, please convert this to a `SerializeFieldAttribute` private field, as per the other places in the codebase. If you need access to this value outside of this type, please use a property instead.

If it does become private, please follow the codebase's private field convention and define the name `_inThisFormat`, as per other places in the codebase.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
+        
+        enemiesRemaining.text = EnemiesAlive.ToString();
+        
+        if (EnemiesAlive > 0)
```
You can one-liner this `if` statement. Nice use of `if` inversions though!

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
+        
     }
 
     IEnumerator SpawnWave()
```
This is missing its accessor. Please provide an accessor.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
+
+        EnemiesAlive = wave.count;
+
+        
```
Nit: needless space in this method. Try to compact it a little more?

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
         {
-            SpawnEnemy();
-            yield return new WaitForSeconds(0.5f);
+            
```
Nit: needless newline.

JohnByrneJames replied:
That looks differnt in the version I have

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
-            yield return new WaitForSeconds(0.5f);
+            
+            SpawnEnemy(wave.enemy);
+            yield return new WaitForSeconds(1f/ wave.rate);
```
Persist the `WaitForSeconds` in a private field and make a new one only when the `wave.rate` changes. This is expensive at the moment.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
     }
 
-    void SpawnEnemy()
+    void SpawnEnemy(GameObject enemy)
```
This is missing its accessor. Please add an accessor.

JohnByrneJames replied:
That is an old version.

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/WaveSpawner.cs:
```diff
@@ -8,50 +8,76 @@
-         
-         Instantiate(enemyPrefab, spawnPoints[spawnPointIndex].position, spawnPoints[spawnPointIndex].rotation);
+        int spawnPointIndex = UnityEngine.Random.Range(0, spawnPoints.Length);
+
```
Nit: needless newline.

JohnByrneJames replied:
Both lines needed (DIFFERENT IN MY VERSION)

RubyNova commented on UnityClient/Assets/Scripts/TowerDefence/Win.cs:
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Win : MonoBehaviour
```
I'm assuming this type is just a placeholder for now, and won't review it as a result this time.

RubyNova reviewed (changes requested):
The code is kind of spaghetti, but if it works then that is good. I've put many comments about refactorings, redesigns and performance bottlenecks.

Also why does this branch have MonoBleedingEdge in it?

JohnByrneJames committed:
Applied changes to branch to make it functional for prototype

Despite some comments being ignored I have made a note and will change them in the future. But that is not important if we are focused on a product for this Tuesday. Last test this worked fine.

JohnByrneJames committed:
Applied changes to branch to make it functional for prototype

Despite some comments being ignored I have made a note and will change them in the future. But that is not important if we are focused on a product for this Tuesday. Last test this worked fine.

JohnByrneJames committed:
Merge remote-tracking branch 'origin/feature/pathfinding' into feature/pathfinding

JohnByrneJames committed:
Merge remote-tracking branch 'origin/feature/pathfinding' into feature/pathfinding

JohnByrneJames committed:
Merge remote-tracking branch 'origin/feature/pathfinding' into feature/pathfinding

JohnByrneJames committed:
Merge remote-tracking branch 'origin/feature/pathfinding' into feature/pathfinding

JohnByrneJames committed:
Merge remote-tracking branch 'origin/feature/pathfinding' into feature/pathfinding

JohnByrneJames committed:
Added Ability to have multiple instances of pathfinding

RubyNova reviewed (approved):
Normally, I'd request changes on code like this but were out of time. It works. Post-prototype the first thing to do will be to clean this up. LGTM otherwise.
